runtime/examples/memtest/memtest_parallel.cpp

```diff
@@ -0,0 +1,164 @@
+
+double calcSpeed(unsigned long amount, unsigned long cycles, float clock) {
+  return (amount / (cycles / (clock * 1000000.0))) / (1024*1024*1024);
+}
+
+void executeRandomBenchmark(tapasco::Tapasco &tapasco, int op, unsigned long cycles, int byte_length, float designclk, int instances) {
+  char text[20];
+  std::vector<tapasco::job_future> jobs;
+  std::vector<tapasco::RetVal<unsigned long>> retvals;
+  unsigned long rets[instances];
+  for (int instance = 0; instance < instances; instance++) {
+    rets[instance] = 0;
+    tapasco::RetVal<unsigned long> ret_val(rets[instance]);
+    retvals.push_back(ret_val);
+    auto job = tapasco.launch(PE_ID, ret_val, 3 + op, cycles, byte_length, 1234, 5678);
+    jobs.push_back(job);
+  }
+  unsigned long acc = 0;
+  for (int instance = 0; instance < instances; instance++) {
+    jobs[instance]();
+    acc += retvals[instance].value;
+    double iops = retvals[instance].value / (cycles / (designclk * 1000000.0));
+    sprintf(text, "%.2fMIOPS", iops / 1000000);
+    std::cout << left << setw(col_width) << setfill(space) << text;
+  }
+  double iops = acc / (cycles / (designclk * 1000000.0));
+  sprintf(text, "%.2fMIOPS", iops / 1000000);
+  std::cout << left << setw(col_width) << setfill(space) << text;
+  sleep(1);
+}
+
+void benchmarkRandom(tapasco::Tapasco &tapasco, float designclk, unsigned long cycles, int instances) {
+  std::cout << "Initializing memory (required for ECC memory)" << std::endl;
+  if (ecc_memory) {
+    auto job = tapasco.launch(PE_ID, 4, 2 * cycles, random_byte_length[random_byte_length_c - 1], 1234, 5678);
+    job();
+    sleep(1);
+    job = tapasco.launch(PE_ID, 4, 2 * cycles, random_byte_length[random_byte_length_c - 1], 5678, 5678);
+    job();
+    sleep(1);
+  }
+  char text[20];
+  std::cout << std::endl << std::endl;
+  std::cout << "Random Access Read/Write (" << cycles << " Cycles)" << std::endl << std::endl;
+
+  std::cout << left << setw(col_width) << setfill(space) << "Size";
+  for (int instance = 0; instance < instances; instance++) {
+    sprintf(text, "Instance %i", instance);
+    std::cout << left << setw(col_width) << setfill(space) << text;
+  }
+  std::cout << left << setw(col_width) << setfill(space) << "Total";
+  std::cout << std::endl;
+
+  for (int count = 0; count < random_byte_length_c; count++) {
+    sprintf(text, "%iB", random_byte_length[count]);
+    std::cout << left << setw(col_width) << setfill(space) << text;
+    executeRandomBenchmark(tapasco, 2, cycles, random_byte_length[count], designclk, instances);
+    std::cout << std::endl;
+  }
+
+}
+
+
+void executeBatchBenchmark(tapasco::Tapasco &tapasco, float designclk, int op, size_t size, int instances) {
+  unsigned long acc = 0;
+  size_t len = 1 << size;
+  unsigned long accs[instances];
+  for (int instance = 0; instance < instances; instance++) accs[instance] = 0;
+  for (int i = 0; i < batch_iterations; i++) {
```

Comment:
Threads? OpenMP might be enough here.

Reply:
At which point exactly do you think multiple threads would be useful? Because the two loops show in this snippet would not benefit from parallelization.

The only point I see where there could be some use are the two inner loops, where the jobs are started (e.g. line 97) and the results are queried (e.g. line 104). But even there I'm not sure if there is measurable impact, because most of the time is spent waiting for the jobs to finish (and not launching the jobs or querying the results)

Reply:
You're right, I wasn't considering the runtime and lack of data transfers.